src/storage.rs

```diff
@@ -0,0 +1,86 @@
+use std::env;
+
+use axum::{response::{IntoResponse, Response}, Json};
+use chrono::Utc;
+use http::StatusCode;
+use serde_json::json;
+use sqlx::{Sqlite, Pool, sqlite::SqlitePoolOptions, Executor, Row};
+
+#[derive(Debug)]
+pub enum StorageError {
+    DatabaseError(sqlx::error::Error)
+}
+
+impl IntoResponse for StorageError {
+    fn into_response(self) -> Response {
+        let message = match self {
+            StorageError::DatabaseError(error) => error.to_string()
+        };
+        let body = Json(json!({ "error": message}));
+        (StatusCode::INTERNAL_SERVER_ERROR, body).into_response()
+    }
+}
+
+#[derive(Clone)]
+pub struct PersistentStorage(Pool<Sqlite>);
+
+impl PersistentStorage {
+    pub async fn has_contributed(&self, uid: &str) -> Result<bool, StorageError> {
+        let sql = "SELECT EXISTS(SELECT 1 FROM contributors WHERE uid = ?1";
+        let result = self.0
+            .fetch_one(sqlx::query(sql).bind(uid))
+            .await
+            .map(|row| row.get(0))
+            .map_err(|e| StorageError::DatabaseError(e));
+
+        result
+    }
+
+    pub async fn insert_contributor(&self, uid: &str) {
+        let sql = "INSERT INTO contributors (uid, started_at) VALUES (?1, ?2)";
+        self.0
+            .execute(sqlx::query(sql).bind(uid).bind(Utc::now()))
+            .await
+            .ok();
+    }
+
+    pub async fn finish_contribution(&self, uid: &str) {
```

Comment:
Do we need this method?

Reply:
I wanted to separate successful contributors (`finish_contribution`) from others (`expire_contribution`). What do you think?

Reply:
Yeah thats fair, I think we could use this information in a meaningful way around the ending of the ceremony